drivers/onewire/onewire.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (C) 2016-2020 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_onewire
+ * @{
+ *
+ * @file
+ * @brief       1-Wire bus driver implementation
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include <string.h>
+
+#include "assert.h"
+#include "xtimer.h"
+#include "onewire.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+#define T_RESET_HOLD        (480)
+#define T_RESET_SAMPLE      (70)
+
+#define T_RW_PULSE          (3)
+#define T_W_HOLD            (57)
+#define T_W_RECOVER         (10)
+#define T_R_SAMPLE          (7)
+#define T_R_RECOVER         (55)
+
+static inline int _read_bit(const onewire_t *owi)
+{
+    gpio_clear(owi->pin);
+    xtimer_usleep(T_RW_PULSE);
+    gpio_init(owi->pin, owi->imode);
+    xtimer_usleep(T_R_SAMPLE);
+    int in = (gpio_read(owi->pin)) ? 1 : 0;
+    xtimer_usleep(T_R_RECOVER);
+    gpio_init(owi->pin, owi->omode);
+    gpio_set(owi->pin);
+
+    return in;
+}
+
+static inline void _write_bit(const onewire_t *owi, int out)
+{
+    gpio_clear(owi->pin);
+    if (out) {
+        /* shift out a `1` */
+        xtimer_usleep(T_RW_PULSE);
+        gpio_set(owi->pin);
+        xtimer_usleep(T_W_HOLD);
+    }
+    else {
+        /* shift out a `0` */
+        xtimer_usleep(T_RW_PULSE + T_W_HOLD);
+        gpio_set(owi->pin);
+    }
+    xtimer_usleep(T_W_RECOVER);
+}
+
+int onewire_init(onewire_t *owi, const onewire_params_t *params)
+{
+    owi->pin = params->pin;
+
+    owi->omode = params->pin_mode;
+    owi->imode = (owi->omode == GPIO_OD_PU) ? GPIO_IN_PU : GPIO_IN;
+
+    if (gpio_init(owi->pin, owi->omode) != 0) {
+        return ONEWIRE_ERR_PINCFG;
+    }
+    gpio_set(owi->pin);
+    return ONEWIRE_OK;
+}
+
+int onewire_reset(const onewire_t *owi, const onewire_rom_t *rom)
+{
+    int res = ONEWIRE_OK;
+
+    gpio_clear(owi->pin);
+    xtimer_usleep(T_RESET_HOLD);
+    gpio_set(owi->pin);
+
+    xtimer_usleep(T_RESET_SAMPLE);
+    gpio_init(owi->pin, owi->imode);
+    if (gpio_read(owi->pin)) {
+        res = ONEWIRE_NODEV;
+    }
+
+    xtimer_usleep(T_RESET_HOLD - T_RESET_SAMPLE);
+    gpio_init(owi->pin, owi->omode);
+    gpio_set(owi->pin);
+
+    if ((res == ONEWIRE_OK) && (rom != NULL)) {
+        onewire_write_byte(owi, ONEWIRE_ROM_MATCH);
+        onewire_write(owi, rom->u8, sizeof(onewire_rom_t));
+    }
+
+    return res;
+}
+
+void onewire_read(const onewire_t *owi, void *data, size_t len)
+{
+    uint8_t *buf = (uint8_t *)data;
```

Comment:
```suggestion
    uint8_t *buf = data;
```

Reply:
works as well, but personally I prefer the explicit cast for readability here.

Reply:
Explicit casts from `void *` are a known anti-pattern. E.g. consider:

```C
int *array = (int *)malloc(sizeof(int) * 42);
```

If the `#include <stdlib.h>` is missing, plain C will treat this as an implicit function declaration of malloc as `int malloc(unsigned long)`. On platforms with `sizeof(int) != sizeof(void *)` this will result in run time errors.

An implicit cast of `int` to `int *` is however even in plain C a reason to warn.

Explicit casts turn of all type safety checks. They should only be used with care and only when really needed.

And honestly, this is more readable without the explicit cast.